### PR TITLE
feat: support untyped NULL value parameters

### PR DESCRIPTION
**Pull request by olavloite**
Support untyped null parameter values for `PreparedStatement`. This will allow users to add null values as a statement parameter value without having to know what type the column that the parameter is bound to uses.

### CODE REVIEW

File: src/main/java/com/google/cloud/spanner/jdbc/AbstractJdbcPreparedStatement.java [hunk 1]
```diff
@@ -81,7 +83,17 @@ public void addBatch(String sql) throws SQLException {
...
+      parameters.setParameter(
+          parameterIndex,
+          Value.untyped(
+              com.google.protobuf.Value.newBuilder().setNullValue(NullValue.NULL_VALUE).build()),
+          sqlType,
+          /* scaleOrLength = */ null);
+    } else {
+      parameters.setParameter(
+          parameterIndex, /* value = */ null, sqlType, /* scaleOrLength = */ null);
+    }
```
Comment (harshachinta) [inline src/main/java/com/google/cloud/spanner/jdbc/AbstractJdbcPreparedStatement.java:96]: Instead of adding this untyped logic here, can we do the following?
1.  Add untyped value binding logic under default case of `setNullValue()` method in `JdbcParameterStore.java`, [Code link](https://github.com/googleapis/java-spanner-jdbc/blob/8f75d26c18844617e275feb952db415a067a91d2/src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java#L928)
2. 
```suggestion
    parameters.setParameter(parameterIndex, null, sqlType, null);
```
3. Also, in [setNullValue()](https://github.com/googleapis/java-spanner-jdbc/blob/8f75d26c18844617e275feb952db415a067a91d2/src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java#L865) method, currently if `sqlType` is null, we bind it to `String` Value, instead we should change it to use below. Otherwise when we do `setObject(1, null)` and if the column is an array type, then we currently bind it to String which are not compatible and backend throws an error. Instead, if we use untyped value then it works for all cases.
```
 binder.to(Value.untyped(com.google.protobuf.Value.newBuilder().setNullValue(NullValue.NULL_VALUE).build());
```
4. Ideally doing this change in [setNullValue()](https://github.com/googleapis/java-spanner-jdbc/blob/8f75d26c18844617e275feb952db415a067a91d2/src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java#L865) method, handles all the following cases by binding untyped null value and also works for other methods like `setObject()`.
-  value is null and type is null
-  value is null and type is Types.OTHER, Types.NULL or not known.

Knut, please correct me if something is unclear or not correct here.

File: src/test/java/com/google/cloud/spanner/jdbc/it/ITJdbcPreparedStatementTest.java [hunk 2]
```diff
@@ -1294,6 +1294,29 @@ public void test12_InsertReturningTestData() throws SQLException {
...
+      try (PreparedStatement preparedStatement =
+          connection.prepareStatement(
+              "insert into all_nullable_types ("
+                  + "ColInt64, ColFloat64, ColBool, ColString, ColBytes, ColDate, ColTimestamp, ColNumeric, ColJson, "
+                  + "ColInt64Array, ColFloat64Array, ColBoolArray, ColStringArray, ColBytesArray, ColDateArray, ColTimestampArray, ColNumericArray, ColJsonArray) "
+                  + "values (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)")) {
+        for (int param = 1;
+            param <= preparedStatement.getParameterMetaData().getParameterCount();
+            param++) {
+          preparedStatement.setNull(param, Types.OTHER);
```
Comment (harshachinta) [inline src/test/java/com/google/cloud/spanner/jdbc/it/ITJdbcPreparedStatementTest.java:1309]: Can we add test for `setObject(param, null)` also? This helps us test the case when sqlType is `null`.

**Review by rajatbhatta — COMMENTED** [review]
JSON and JSONB are nullable. But I see that the [setNullValue()](https://github.com/googleapis/java-spanner-jdbc/blob/8f75d26c18844617e275feb952db415a067a91d2/src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java#L869) method does not cover JSON and JSONB data types (JsonType.VENDOR_TYPE_NUMBER and PgJsonbType.VENDOR_TYPE_NUMBER are missing from switch case). Can we add these too and verify once if all supported data types are included in the switch case so we do not miss out on any case? Let me know if we should do this as a separate exercise.

File: src/test/java/com/google/cloud/spanner/jdbc/it/ITJdbcPreparedStatementTest.java (hunk 2 above)
Comment (olavloite) [reply src/test/java/com/google/cloud/spanner/jdbc/it/ITJdbcPreparedStatementTest.java:1309]: Yes, good point, done.

**Comment by olavloite** [comment]
> JSON and JSONB are nullable. But I see that the [setNullValue()](https://github.com/googleapis/java-spanner-jdbc/blob/8f75d26c18844617e275feb952db415a067a91d2/src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java#L869) method does not cover JSON and JSONB data types (JsonType.VENDOR_TYPE_NUMBER and PgJsonbType.VENDOR_TYPE_NUMBER are missing from switch case). Can we add these too and verify once if all supported data types are included in the switch case so we do not miss out on any case? Let me know if we should do this as a separate exercise.

Good point, I've added those to the switch case as well.

**Review by rajatbhatta — APPROVED** [review]
LGTM

File: src/main/java/com/google/cloud/spanner/jdbc/AbstractJdbcPreparedStatement.java (hunk 1 above)
Comment (harshachinta) [reply src/main/java/com/google/cloud/spanner/jdbc/AbstractJdbcPreparedStatement.java]: Thanks, Knut.

**Review by harshachinta — APPROVED** [review]
LGTM
